File: 3_bruteforce_backtracking/강대훈/전력망을 둘로 나누기.swift
import Foundation

func solution(_ n:Int, _ wires:[[Int]]) -> Int {
    var graph = [[Int]](repeating: [], count: n + 1)
    var ans = Int.max

    for wire in wires {
        graph[wire[0]].append(wire[1])
        graph[wire[1]].append(wire[0])
    }

    for wire in wires {
        let (v1, v2) = (wire[0], wire[1])
        var visited = [Bool](repeating: false, count: n + 1)
        var cnt = 0

        visited[v1] = true
        visited[v2] = true

        dfs(v1)

        ans = min(ans, abs(cnt - (n - cnt)))

        func dfs(_ n: Int) {
            cnt += 1

            for next in graph[n] {
                if visited[next] { continue }
                visited[next] = true
                dfs(next)
            }
        }
    }

    return ans
}

// 1 : [2]
// 2 : [1, 7]
// 3 : [4, 7]
// 4 : [3, 5]
// 5 : [4]
// 6 : [7]
// 7 : [2, 3, 6]

File: 3_bruteforce_backtracking/김민주/전력망을둘로나누기.java
import java.util.*;

class Solution {
    int na, nb;
    int mn;
    List <Integer>[] arr;
    int cnt;
    int dfs(int node, boolean[] visited){
        visited[node] = true;
        cnt += 1;
        
        for (int newone: arr[node]){
            if (!visited[newone]){
                dfs(newone, visited);
                
            }
        }
        return cnt;
        
    }
    
    public int solution(int n, int[][] wires) {
        int ln = wires.length;
        int answer = ln;
        int cntDfs = ln;
        arr = new ArrayList[n+1];
        mn = n;
        
        for (int i=1; i<n+1; i++){
            arr[i] = new ArrayList<>();
        }
        for (int i=0; i<ln; i++){
            arr[wires[i][0]].add(wires[i][1]);
            arr[wires[i][1]].add(wires[i][0]);
        }
        
        for (int i=0; i<ln; i++){
            int nowi = wires[i][0];
            int nowj = wires[i][1];
            arr[nowi].remove(Integer.valueOf(nowj));
            arr[nowj].remove(Integer.valueOf(nowi));
            cnt = 0;
            boolean[] visited = new boolean [n+1];

            cntDfs = dfs(wires[0][0], visited);
            int a = Math.min(n-cntDfs, cntDfs);
            int b = Math.max(n-cntDfs, cntDfs);
            answer = Math.min(b-a, answer);
            
            arr[nowi].add(nowj);
            arr[nowj].add(nowi);
            
        }
        return answer;
    }
}